### PR TITLE
Google AI Gemini: replace OkHttp and Retrofit with Java 11 HttpClient

### PR DESCRIPTION
**Pull request by Bjarne-Kinkel**
## Issue
Based on #1903

## Change
Replaced OkHttp and Retrofit inside the GeminiService with an implementation using the HttpClient (Java 11).

## General checklist

- [X] There are no breaking changes
- [X]  I have added unit and integration tests for my change (already existing)
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java [hunk 1]
```diff
@@ -1,85 +1,94 @@
 package dev.langchain4j.model.googleai;
 
-//import io.reactivex.rxjava3.core.Observable;
-import okhttp3.OkHttpClient;
-import okhttp3.logging.HttpLoggingInterceptor;
+import com.google.gson.Gson;
```
Comment (kpavlov) [inline langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java:3]: There is a ticket [#1810](https://github.com/langchain4j/langchain4j/issues/1810) to migrate from Gson to Jackson. Should Gson be introduced here?

**Review by kpavlov — COMMENTED** [review]
Thank you, @Bjarne-Kinkel

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java (hunk 1 above)
Comment (Bjarne-Kinkel) [reply langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java:3]: Up until now, only Gson has been used in this module, but sure, I can switch to Jackson as well.

Comment (Bjarne-Kinkel) [reply langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java:3]: When using Jackson, each (sub)class will require a dedicated no-argument constructor due to the use of the @Builder annotation within the POJOs. Should I continue, or would it be better to create a separate issue for that?

Comment (langchain4j) [reply langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java:3]: @Bjarne-Kinkel let's keep it separate then

**Comment by Bjarne-Kinkel** [comment]
I ran the tests today, and they all passed. Is there anything else that needs to be done or improved?

**Comment by langchain4j** [comment]
@Bjarne-Kinkel thank you! I will try to review it today or tomorrow

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java [hunk 2]
```diff
@@ -1,85 +1,94 @@
...
+        HttpRequest request = buildHttpRequest(url, apiKey, jsonBody);
 
-            clientBuilder.addInterceptor(logging);
+        logRequest(jsonBody);
+
+        try {
+            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+
+            if (response.statusCode() >= 300) {
+                throw new IOException(String.format("HTTP error (%d): %s", response.statusCode(), response.body()));
```
Comment (langchain4j) [inline langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java]: It was throwing `RuntimeException` before, is there a reason for this change?

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java [hunk 3]
```diff
@@ -1,85 +1,94 @@
...
+            }
+
+            logResponse(response.body());
+
+            return gson.fromJson(response.body(), responseType);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Sending request was interrupted", e);
```
Comment (langchain4j) [inline langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java]: Is it correct to throw `IOException` here?

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java [hunk 4]
```diff
@@ -1,85 +1,94 @@
...
-        if (logger != null) {
-            HttpLoggingInterceptor logging = new HttpLoggingInterceptor(logger::debug);
-            logging.redactHeader(API_KEY_HEADER_NAME);
-            logging.setLevel(HttpLoggingInterceptor.Level.BODY);
+        this.httpClient = HttpClient.newBuilder()
+                .connectTimeout(timeout)
+                .build();
+    }
+
+    public GeminiGenerateContentResponse generateContent(String modelName, String apiKey, GeminiGenerateContentRequest request) throws IOException {
```
Comment (langchain4j) [inline langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GeminiService.java]:
```suggestion
    GeminiGenerateContentResponse generateContent(String modelName, String apiKey, GeminiGenerateContentRequest request) throws IOException {
```
nit: please remove `public` from the members of this class

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java [hunk 5]
```diff
@@ -64,31 +59,17 @@ public GoogleAiEmbeddingModel(
...
-                    GeminiError error = GSON.fromJson(errorBody.string(), GeminiErrorContainer.class).getError();
-
-                    throw new RuntimeException(
-                        String.format("%s (code %d) %s", error.getStatus(), error.getCode(), error.getMessage()));
-                }
-            }
-        } catch (IOException e) {
-
+            geminiResponse = withRetry(() -> this.geminiService.embed(this.modelName, this.apiKey, embeddingRequest), this.maxRetries);
+        } catch (RuntimeException e) {
```
Comment (langchain4j) [inline langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java]: Is this `try-catch` and wrapping `RuntimeException` into another `RuntimeException` required?

Worth checking all other places as well.

**Review by ghost — COMMENTED** [review]
@Bjarne-Kinkel thank you!

File: langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java (hunk 5 above)
Comment (Bjarne-Kinkel) [reply langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java]: Different RuntimeException messages were thrown depending on the endpoint called (e.g., token count, API, embed). The try catch was necessary to maintain the current functionality, though I'm unsure if it's truly required. Removing the exceptions would result in less nested exceptions.

Comment (Bjarne-Kinkel) [reply langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java]: I've removed it. Feel free to check out the updated version and let me know if it's okay.

**Review by ghost — APPROVED** [review]
@Bjarne-Kinkel thank you!

**Pull request merged**